In [30]:
import pandas as pd

#Reading files
products = pd.read_csv("PRODUCTS_TAKEHOME.csv")
transactions = pd.read_csv("TRANSACTION_TAKEHOME.csv")
users= pd.read_csv("USER_TAKEHOME.csv")


products_info = products.info()
products_info


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 845552 entries, 0 to 845551
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   CATEGORY_1    845441 non-null  object 
 1   CATEGORY_2    844128 non-null  object 
 2   CATEGORY_3    784986 non-null  object 
 3   CATEGORY_4    67459 non-null   object 
 4   MANUFACTURER  619078 non-null  object 
 5   BRAND         619080 non-null  object 
 6   BARCODE       841527 non-null  float64
dtypes: float64(1), object(6)
memory usage: 45.2+ MB


In [8]:
products.head()

CATEGORY_1              CATEGORY_2                   CATEGORY_3  \
0  Health & Wellness           Sexual Health  Conductivity Gels & Lotions   
1             Snacks           Puffed Snacks         Cheese Curls & Puffs   
2  Health & Wellness               Hair Care        Hair Care Accessories   
3  Health & Wellness               Oral Care                   Toothpaste   
4  Health & Wellness  Medicines & Treatments               Essential Oils   

  CATEGORY_4                                       MANUFACTURER  \
0        NaN                                                NaN   
1        NaN                                                NaN   
2        NaN                           PLACEHOLDER MANUFACTURER   
3        NaN                                  COLGATE-PALMOLIVE   
4        NaN  MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...   

             BRAND       BARCODE  
0              NaN  7.964944e+11  
1              NaN  2.327801e+10  
2          ELECSOP  4.618178e+11  
3          COLGATE  3.500047e+10  
4  MAPLE HOLISTICS  8.068109e+11

In [9]:
transactions_info = transactions.info()
transactions_info


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   RECEIPT_ID      50000 non-null  object 
 1   PURCHASE_DATE   50000 non-null  object 
 2   SCAN_DATE       50000 non-null  object 
 3   STORE_NAME      50000 non-null  object 
 4   USER_ID         50000 non-null  object 
 5   BARCODE         44238 non-null  float64
 6   FINAL_QUANTITY  50000 non-null  object 
 7   FINAL_SALE      50000 non-null  object 
dtypes: float64(1), object(7)
memory usage: 3.1+ MB


In [10]:
transactions.head()

RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
4  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

        BARCODE FINAL_QUANTITY FINAL_SALE  
0  1.530001e+10           1.00             
1           NaN           zero       1.49  
2  7.874223e+10           1.00             
3  7.833997e+11           zero       3.49  
4  4.790050e+10           1.00

In [6]:
users_info = users.info()
users_info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   ID            100000 non-null  object
 1   CREATED_DATE  100000 non-null  object
 2   BIRTH_DATE    96325 non-null   object
 3   STATE         95188 non-null   object
 4   LANGUAGE      69492 non-null   object
 5   GENDER        94108 non-null   object
dtypes: object(6)
memory usage: 4.6+ MB


In [11]:
users.head()

ID               CREATED_DATE  \
0  5ef3b4f17053ab141787697d  2020-06-24 20:17:54.000 Z   
1  5ff220d383fcfc12622b96bc  2021-01-03 19:53:55.000 Z   
2  6477950aa55bb77a0e27ee10  2023-05-31 18:42:18.000 Z   
3  658a306e99b40f103b63ccf8  2023-12-26 01:46:22.000 Z   
4  653cf5d6a225ea102b7ecdc2  2023-10-28 11:51:50.000 Z   

                  BIRTH_DATE STATE LANGUAGE  GENDER  
0  2000-08-11 00:00:00.000 Z    CA   es-419  female  
1  2001-09-24 04:00:00.000 Z    PA       en  female  
2  1994-10-28 00:00:00.000 Z    FL   es-419  female  
3                        NaN    NC       en     NaN  
4  1972-03-19 00:00:00.000 Z    PA       en  female

In [14]:
# Checking for missing value in products
missing_values_products = products.isnull().sum()
missing_values_products


CATEGORY_1         111
CATEGORY_2        1424
CATEGORY_3       60566
CATEGORY_4      778093
MANUFACTURER    226474
BRAND           226472
BARCODE           4025
dtype: int64

In [16]:
# Checking for missing value in transactions
missing_values_transactions = transactions.isnull().sum()
missing_values_transactions

RECEIPT_ID           0
PURCHASE_DATE        0
SCAN_DATE            0
STORE_NAME           0
USER_ID              0
BARCODE           5762
FINAL_QUANTITY       0
FINAL_SALE           0
dtype: int64

In [18]:
# Checking for missing value in users
missing_values_users = users.isnull().sum()
missing_values_users

ID                  0
CREATED_DATE        0
BIRTH_DATE       3675
STATE            4812
LANGUAGE        30508
GENDER           5892
dtype: int64

In [20]:
## Checking for duplicates in product
duplicate_rows_products = products.duplicated().sum()
duplicate_rows_products


np.int64(215)

In [23]:
## Checking for duplicates in transactions
duplicate_rows_transactions = transactions.duplicated().sum()
duplicate_rows_transactions


np.int64(171)

In [24]:
## Checking for duplicates in users
duplicate_rows_users = users.duplicated().sum()
duplicate_rows_users

np.int64(0)

In [28]:
# Checking for unique values in products categorical fields to identify challenging columns
unique_values_products = products.nunique()
unique_values_products


CATEGORY_1          27
CATEGORY_2         121
CATEGORY_3         344
CATEGORY_4         127
MANUFACTURER      4354
BRAND             8122
BARCODE         841342
dtype: int64

In [29]:
# Checking for unique values in transactions categorical fields to identify challenging columns
unique_values_transactions = transactions.nunique()
unique_values_transactions


RECEIPT_ID        24440
PURCHASE_DATE        89
SCAN_DATE         24440
STORE_NAME          954
USER_ID           17694
BARCODE           11027
FINAL_QUANTITY       87
FINAL_SALE         1435
dtype: int64

In [31]:
# Checking for unique values in users categorical fields to identify challenging columns
unique_values_users = users.nunique()
unique_values_users

ID              100000
CREATED_DATE     99942
BIRTH_DATE       54721
STATE               52
LANGUAGE             2
GENDER              11
dtype: int64

Missing Values:

Products Dataset:

CATEGORY_1 = (111 missing)



CATEGORY_2 = (1,424 missing)


CATEGORY_3 = (60,566 missing)


CATEGORY_4 = (778,093 missing) – Highly incomplete


MANUFACTURER = (226,474 missing) – lot of  missing data


BRAND = (226,472 missing) – lot of missing data


BARCODE = (4,025 missing)

Transactions Dataset:

BARCODE = (5,762 missing) – Potential issue in linking transactions to products


Users Dataset:

BIRTH_DATE = (3,675 missing)

STATE = (4,812 missing)

LANGUAGE = (30,508 missing) – Significant missing data

GENDER = (5,892 missing)

#Duplicate Rows:

Products Dataset: 215 duplicate rows

Transactions Dataset: 171 duplicate rows

Users Dataset: No duplicates found











#Fields that are challenging to understand:

Products Dataset:

CATEGORY_4 has only 67,459 non-null values in a dataset of 845,552 rows. This suggests that it is either sparsely populated or inconsistently used.

MANUFACTURER and BRAND have a large number of missing values, making it difficult to categorize products.


Transactions Dataset:

FINAL_QUANTITY has entries like "zero" instead of numeric values.

FINAL_SALE may need further normalization, as it is stored as a string rather than a numeric value.

Users Dataset:

GENDER has 11 unique values, which suggests inconsistent data entry (e.g., different labels for the same gender).

LANGUAGE has only 2 unique values, meaning it may not be a useful field if most users have the same language preference.